### ISSUE TITLE
Image "fill" mode?

### ISSUE THREAD

**Issue by brandoncc** [prawnpdf/prawn#1110]
Is there a way to use a "fill" mode instead of "fit" mode for images? I found https://stackoverflow.com/a/30897942/617243 but I am unsure about using it since it uses the internal API and might not even work anymore since it is over 3 years old.

I'm open to a bounding box that clips, or any other implementation that could happen.

Thanks.

**Comment by gettalong** [comment]
I gather that with "fill" you want to have an image fill a certain area? If so, this is easy to do with just the `#image` method and the `width` and `height` arguments:

~~~ ruby
require 'prawn'

Prawn::Document.generate('image.pdf') do |doc|
  doc.image(ARGV[0], width: 200, height: 50)
end
~~~

This will put the image specified on the command line at the current cursor position and it will fill an area of the size 200x50. You could also use `doc.bounds` to fill the current bounding box.

**Comment by brandoncc** [comment]
Thanks for your response. Will that approach maintain the correct image ratio via clipping? That is how "fill" generally works, but the command you shared looks like it will probably stretch/squash images 🤔

**Comment by gettalong** [comment]
Yes, the command will just fill the area with the image, ignoring the image ratio.

So I gather that "fill" for you means filling the whole specified area by respecting the image ratio and clipping the outside?

Then you have to do some parts manually:

~~~ ruby
require 'prawn'

Prawn::Document.generate('image.pdf') do |doc|
  box_w, box_h = 200, 50
  doc.save_graphics_state do
    doc.rectangle([0, doc.cursor], box_w, box_h)
    doc.renderer.add_content(" W n ")
    doc.image(ARGV[0], width: box_w)
  end
end
~~~

What this does is it defines an area via the `#rectangle` method and manual PDF operators where everything outside gets clipped, i.e. not shown. Then the image is drawn with either `width` or `height` (you have to determine yourself which one to use based on the box and image ratios). Thus the image retains its ratio, fills the box and everything outside is clipped.

Note: You have to wrap the operation in `#save_graphics_state` as the clipping operation is non-reversible.

**Comment by brandoncc** [comment]
Very helpful, I love the explanation you provided for what is actually happening. Thank you for taking the time to respond.
